include/accessor.h

```diff
@@ -254,12 +254,17 @@ class accessor<DataT, Dims, Mode, target::device> : public detail::accessor_base
 	}
 
 	template <access_mode M = Mode>
-	inline std::enable_if_t<detail::access::mode_traits::is_producer(M), DataT&> operator[](const id<Dims>& index) const {
-		return m_device_ptr[get_linear_offset(index)];
-	}
-
-	template <access_mode M = Mode>
-	inline std::enable_if_t<detail::access::mode_traits::is_pure_consumer(M), const DataT&> operator[](const id<Dims>& index) const {
+	inline std::conditional_t<detail::access::mode_traits::is_producer(M), DataT&, const DataT&> operator[](const id<Dims>& index) const {
+#if CELERITY_ACCESSOR_BOUNDARY_CHECK
+		if(((index - m_backing_buffer_offset) >= m_backing_buffer_range) != detail::id_cast<Dims>(id<3>(false, false, false))) {
+			for(int d = 0; d < Dims; ++d) {
+				sycl::atomic_ref<size_t, sycl::memory_order::relaxed, sycl::memory_scope::device>{m_attempted_index[0][d]}.fetch_min(index[d]);
+				sycl::atomic_ref<size_t, sycl::memory_order::relaxed, sycl::memory_scope::device>{m_attempted_index[1][d]}.fetch_max(index[d] + 1);
+			}
+			// TODO: make sure we always allocate at least one element in the backing buffer, if CELERITY_ACCESSOR_BOUNDARY_CHECK is enabled
+			return m_device_ptr[0];
```

Comment:
the reasoning here was to prevent the program from crashing (as it would if it were to run in Release build instead of Debug) and to give a cmdln error message to the user, warning that their data is incorrect because of the access. 
Also in some cases doing the out-of-bounds access would result in a hipsycl runtime crash with no info about why is the program crashing, that is why I opted for this instead.

Reply:
I still think out-of-bounds writes shouldn't manifest anywhere in a buffer, that's confusing. A better way of going about this might be to have a "catch all" member of type `DataT` within each accessor, to which a reference is returned in this case.

Reply:
as a sort of default "there was a mistake" value of DataT ? but that would have to be provided either when constructing the `accessor` or the `buffer`, and either options seems more cumbersome. Alternatively, I would be willing to simply attempt the out-of-bounds access, provided @fknorr and @PeterTh agree on it too.

Reply:
No, just a default-constructed value of type `DataT`, simply because we need *something* to return a reference to from `operator[]`.